docs/topics/proc_exposing-application-metrics-using-prometheus-with-vertx.adoc

```diff
@@ -0,0 +1,189 @@
+* Collect and view the data using Prometheus.
+
+Prometheus connects to a monitored application to collect data; the application does not send metrics to a server.
+
+.Prerequisites
+
+* Prometheus configured to collect metrics from the application:
++
+--
+. Download and extract the link:https://prometheus.io/download/[archive^] with the latest Prometheus release:
++
+[source,bash,options="nowrap"]
+----
+$ wget https://github.com/prometheus/prometheus/releases/download/v2.4.3/prometheus-2.4.3.linux-amd64.tar.gz
+$ tar -xvf  prometheus-2.4.3.linux-amd64.tar.gz
+----
+
+. Navigate to the directory with Prometheus:
++
+[source,bash,options="nowrap"]
+----
+$ cd  prometheus-2.4.3.linux-amd64
+----
+
+. Append the following snippet to the `prometheus.yml` file to make Prometheus automatically collect metrics from your application:
++
+[source,yaml,options="nowrap"]
+----
+  scrape_configs:
+    - job_name: 'prometheus'
+      static_configs:
+      - targets: ['localhost:9090']
+    - job_name: 'vertx-app'
+      static_configs:
+      - targets: ['localhost:8081']
+----
++
+The default behavior of {Vertx}-based applications is to expose metrics at the `/metrics` endpoint.
+This is what the MicroProfile Metrics specification requires, and also what Prometheus expects.
+--
+
+* The Prometheus server started on `localhost`:
++
+--
+Start Prometheus and wait until the `Server is ready to receive web requests` message is displayed in the console.
+
+[source,bash,options="nowrap"]
+----
+$ ./prometheus
+----
+--
+
+
+.Procedure
+
+. Include the `vertx-micrometer`, `micrometer-registry-prometheus` and `vertx-web` dependencies in the `pom.xml` file of your application:
++
+.pom.xml
+[source,xml]
+----
+<dependency>
+  <groupId>io.vertx</groupId>
+  <artifactId>vertx-micrometer-metrics</artifactId>
+  <version>3.5.4</version>
+</dependency>
+<dependency>
+  <groupId>io.micrometer</groupId>
+  <artifactId>micrometer-registry-prometheus</artifactId>
+  <version>1.0.0</version>
+</dependency>
+<dependency>
+  <groupId>io.vertx</groupId>
+  <artifactId>vertx-web</artifactId>
+  <version>3.5.4</version>
+</dependency>
+----
+
+. Annotate methods or classes with the metrics annotations, for example:
++
+--
+[source,java]
+----
+package org.acme;
+
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.config.MeterFilter;
+import io.micrometer.core.instrument.distribution.DistributionStatisticConfig;
+import io.micrometer.prometheus.PrometheusMeterRegistry;
+import io.vertx.core.AbstractVerticle;
+import io.vertx.core.Vertx;
+import io.vertx.core.VertxOptions;
+import io.vertx.core.http.HttpServerOptions;
+import io.vertx.micrometer.MicrometerMetricsOptions;
+import io.vertx.micrometer.VertxPrometheusOptions;
+import io.vertx.micrometer.backends.BackendRegistries;
+
+public class MyVertxApp extends AbstractVerticle {
+
+    public static void main(String[] args) {
+        Vertx vertx = Vertx.vertx(new VertxOptions().setMetricsOptions(
```

Comment:
Since you use the Vert.x Maven plugin (mvn vertx:run) then you can't define options like this. You must create your own Launcher class and override the beforeStartingVertx method. Then in the POM file, declare your launcher class.

Reply:
Is there a related documented where we describe how to setup the project? In particular usage of the Vert.x Maven plugin?

Reply:
Well, either create `Launcher` class and override the method, or you should also be able make it work by doing following:
- move everything from `public static void main` to `public void start()` method, just before `timer` definition
- set up `<vertx.verticle>` property to the name of a class (in this case, `org.acme.MyVertxApp`) in project's `pom.xml` properties section.
- define `vertx-maven-plugin` properly, in project's `pom.xml`

And then, if I am hopefully right, you should be able to execute `mvn vertx:run`. Here is the [VMP documentation](https://reactiverse.io/vertx-maven-plugin/) for reference.

Reply:
After consulting with @spisiakm, the last bullet only means that the configuration should be present in the documentation, which it is (see [Configuring your application to use Eclipse Vert.x, step 2](https://launcher.fabric8.io/docs/vertx-runtime.html#configuring-your-application-to-use-vertx_vertx)).

Reply:
Thanks for the comments Martin. I think I attempted to address your first point, hopefully correctly. I need a little more help for your second comment as I don't the correct syntax to specify the dependency. I think it should be added to the module. Regarding your last point and Sheldon's comment, I can add a link to the if you feel it is necessary but I do think that the configuration is already documented.